:memo: Rework access site page and add file transfer page (#2649)
platformsh/platformsh-docs: docs/src/development/file-transfer.md

---
title: "Transfer files to and from your app"
weight: 7
sidebarTitle: Transfer files
---

After your app is built, its file system is read-only.

> This means that the only way you can edit your app's files and code is through Git.

Reviewer: :hammer_and_wrench: A lot of files aren't managed through Git, so I'd take that part out. Especially as it's somewhat contradicted by the next sentence.

```suggestion
This means that the only way you can edit your app's code is through Git.
```

However, you can transfer files to and from your built app without using Git.

> To do so, you need to configure [mounts](../create-apps/app-reference.md#mounts).

Reviewer: :hammer_and_wrench:  This seems to imply that both to and from require mounts because the `so` here seems to cover both. I think this change might solve that? But could be something better :shrug: 
```suggestion
To transfer files to the app, you need to configure [mounts](../create-apps/app-reference.md#mounts).
```

Author: See new introduction rework for improved clarity.

Mounts let you set up directories that remain writable after the build is complete.
You can then transfer files directly to and from mounts inside your app
with a single command via the [Platform.sh CLI](../administration/cli/_index.md).

Alternatively, you can transfer files to and from your built app 

> using an SSH client such `scp` and `rsync`.

Reviewer: :+1: This part seems clear enough that it's possible throughout the app without thinking about mounts.

> such as `scp` or `rsync`.

Author: Reworked the introduction to make it clear from the get-go that 2 options are available to upload and download files to/from a built app:

- Mounts (cf. `mount:upload` and `mount:download` commands).
- SSH clients (now we have upload and download commands documented for both `scp` and `rsync`).

reverse the order of the parameters:

```bash
scp diagram.png "$(platform ssh --pipe -p {{< variable "PROJECT_ID" >}} -e {{< variable "ENVIRONMENT_NAME" >}})":web/uploads
```

For other options, see the [`scp` documentation](https://www.man7.org/linux/man-pages/man1/scp.1.html).

### rsync

You can use `rsync` to copy files to and from a remote environment.

For example, to copy all the files in the `web/uploads` directory on the remote environment 
to the local `uploads` directory,
run the following command 
(replacing {{< variable "PROJECT_ID" >}} and {{< variable "ENVIRONMENT_NAME" >}} with appropriate values):

```bash
rsync -az "$(platform ssh --pipe -p {{< variable "PROJECT_ID" >}} -e {{< variable "ENVIRONMENT_NAME" >}})":web/uploads/ ./uploads/
```

To copy files from your local directory to the Platform.sh environment, 
reverse the order of the parameters:

```bash
rsync -az uploads/ "$(platform ssh --pipe -p {{< variable "PROJECT_ID" >}} -e {{< variable "ENVIRONMENT_NAME" >}})":web/uploads/
```

Note that `rsync` is very sensitive about trailing `/` characters.

If you're using UTF-8 encoded files on macOS, 

> add the `--iconv=utf-8-mac,utf-8` flag to your `rsync`call.

Reviewer: :clipboard: Missing space
```suggestion
add the `--iconv=utf-8-mac,utf-8` flag to your `rsync` call.
```

---
title: "Transfer files to and from your app"
weight: 7
sidebarTitle: File Transfer
---

After your app is built, its file system is read-only. 
This means that the only way you can edit your app's code is through Git.

However, you can transfer files to and from your app through mounts.
To do so, follow these steps.

## Before you begin

You need to [set up directories that remain writable after your app's build is complete](../create-apps/app-reference.md#mounts).

> These are known as mounts.

Author: Done!

Are you sure you want to continue? [Y/n]
Downloading files from the remote mount /app/private to /Users/alice/Projects/foo/private
  receiving file list ...
 done

  sent 16 bytes  received 3.73K bytes  2.50K bytes/sec
  total size is 1.77M  speedup is 471.78
  time: 0.91s
The download completed successfully.
```

Similarly, to upload the files contained in the local `private` directory to the `private` mount,
run the following command: 

```bash
$ platform mount:upload --mount private --source ./private
```

You get the following output:

```bash
This will add, replace, and delete files in the remote mount 'private'.

Are you sure you want to continue? [Y/n]
Uploading files from /Users/alice/Projects/foo/private to the remote mount /app/private
  building file list ...
 done

  sent 2.35K bytes  received 20 bytes  1.58K bytes/sec
  total size is 1.77M  speedup is 745.09
  time: 0.72s
The upload completed successfully.
```

## Transfer files using an SSH client

Another way to transfer files to and from a mount is to use an SSH client. 

To do so, you need to connect to your app with SSH.
Follow these steps:

> 1. Optional: If you've never done it before, [generate an SSH key and add it to your Platform account](../development/ssh/ssh-keys.md#add-ssh-keys).

Author: Section removed as unneeded.

(replacing `<PROJECT_ID>` and `<ENVIRONMENT_NAME>` with appropriate values):

```bash
scp "$(platform ssh --pipe -p <PROJECT_ID> -e <ENVIRONMENT_NAME>)":web/uploads/diagram.png .
```

The `diagram.png` file is copied to the current local directory.

To copy files from your local directory to the Platform.sh environment, 
reverse the order of the parameters:

```bash
scp diagram.png "$(platform ssh --pipe -p <PROJECT_ID> -e <ENVIRONMENT_NAME>)":web/uploads
```

See the [scp documentation](https://www.man7.org/linux/man-pages/man1/scp.1.html) for other options.

### rsync

`rync` is another tool you can use to copy files to and from a remote environment.

For example, to copy all the files in the `web/uploads` directory on the remote environment 
to the local `uploads` directory,
run the following command 
(replacing `<PROJECT_ID>` and `<ENVIRONMENT_NAME>` with appropriate values):

```bash
rsync -az "$(platform ssh --pipe -p <PROJECT_ID> -e <ENVIRONMENT_NAME>)":web/uploads/ ./uploads/
```

Note that rsync is very sensitive about trailing `/` characters.
Consult the [rsync documentation](https://man7.org/linux/man-pages/man1/rsync.1.html) for more details.

> See the guides on [migrating](../tutorials/migrating.md) and [exporting](../tutorials/exporting.md) for more examples using `rsync`.

Author: I have removed this sentence, and also removed the duplicate information from the Downloading files section in the Exporting tutorial and linked to the Transfer files page instead as suggested.

sidebarTitle: File Transfer
---

After your app is built, its file system is read-only.

The only way you can transfer files to your built app without using Git is through [mounts](../create-apps/app-reference.md#mounts).
Mounts let you set up directories that remain writable after the build is complete.
You can then upload files directly to mounts inside your app.

To download files from your built app, 

> you can also use mounts or [an alternative data export method](../tutorials/exporting.md).

Author: Actually, there isn't much information on alternative methods to download files from a built app in the Exporting data tutorial (apart from the `project:get` command that downloads the whole project). Are there any other specific ways you were thinking of and that I should explicitly mention or link to?

Author: Introduction reworked
